src/sst/core/interprocess/mmapchild_pin3.h

```diff
@@ -0,0 +1,114 @@
+// Copyright 2009-2019 NTESS. Under the terms
+// of Contract DE-NA0003525 with NTESS, the U.S.
+// Government retains certain rights in this software.
+//
+// Copyright (c) 2009-2019, NTESS
+// All rights reserved.
+//
+// This file is part of the SST software package. For license
+// information, see the LICENSE file in the top level directory of the
+// distribution.
+
+#ifndef SST_CORE_INTERPROCESS_TUNNEL_MMAP_CHILD_PIN3_H
+#define SST_CORE_INTERPROCESS_TUNNEL_MMAP_CHILD_PIN3_H 1
+
+#include "pin.H"
+#include <iostream>
+#include <fstream>
+#include <sys/shm.h>
+#include <sys/stat.h>
+#include "atomic.hpp"
+
+#include "sst/core/interprocess/tunneldef.h"
+
+namespace SST {
+namespace Core {
+namespace Interprocess {
+
+
+/** Class supports an IPC tunnel between two or more processes, via an mmap'd file
+ * This class attaches to an existing tunnel for a child process using PinCRT
+ *
+ * @tparam TunnelType Tunnel definition
+ */
+template<typename TunnelType>
+class MMAPChild_Pin3 {
+
+public:
+    /** Child-side tunnel manager for an IPC tunnel
+     * Opens an existing file and mmaps it using PinCRT
+     *
+     * @param file_name Name of the shared file to mmap
+     */
+    MMAPChild_Pin3(const std::string& file_name)
+    {
+        filename = file_name;
+        NATIVE_FD fd;
+        OS_RETURN_CODE retval = OS_OpenFD(filename.c_str(), OS_FILE_OPEN_TYPE_READ | OS_FILE_OPEN_TYPE_WRITE, 0, &fd);
+
+        if (!OS_RETURN_CODE_IS_SUCCESS(retval)) {
+            // Not using Output because IPC means Output might not be available
+            fprintf(stderr, "Failed to open file for IPC '%s': %s\n",
+                    filename.c_str(), strerror(errno));
+            exit(1);
+        }
+
+        shmPtr = NULL;
+        retval = OS_MapFileToMemory(NATIVE_PID_CURRENT, OS_PAGE_PROTECTION_TYPE_READ | OS_PAGE_PROTECTION_TYPE_WRITE, 
+                sizeof(InternalSharedData), OS_MEMORY_FLAGS_SHARED, fd, 0, &shmPtr);
+
+        if (!OS_RETURN_CODE_IS_SUCCESS(retval)) {
+            // Not using Output because IPC means Output might not be available
+            fprintf(stderr, "mmap failed\n");
```

Comment:
Do the return codes mean anything? If so, we could be be better about printing `strerror`

Reply:
I'm not sure how detailed the return codes are, the documentation for this function says it can return OS_RETURN_CODE_NO_ERROR or OS_RETURN_CODE_MEMORY_MAP_FAILED but I don't know if it provides more detail than that.

Reply:
Ha, okay, so this really is a binary pass/no pass

Reply:
The return value structure also captures errno, so I added better error printing.